File: apps/web/lib/api/analytics.ts
"use client";

/**
 * Owner / admin analytics overview client.
 *
 * Why one shape, not many
 * -----------------------
 * The backend returns every analytics block in a single
 * ``/v1/analytics/overview`` payload — see
 * ``apps/api/shiftops_api/api/v1/analytics.py`` for the rationale. We
 * mirror that here so the UI does one fetch and renders many cards from
 * the same store of truth. ``previous`` is a recursive payload of the
 * same shape (without its own ``previous``), filled when ``compare`` is
 * requested.
 */

import { api, type ApiResult } from "@/lib/api/client";

export type DensityFlag = "ok" | "low" | "empty";

export interface AnalyticsKpi {
  shiftsClosed: number;
  shiftsClean: number;
  shiftsWithViolations: number;
  averageScore: number | null;
  cleanlinessRate: number | null;
}

export interface AnalyticsHeatmapCell {
  dayOfWeek: number;
  hourOfDay: number;
  shiftCount: number;
  averageScore: number | null;
}

export interface AnalyticsViolator {
  userId: string;
  fullName: string;
  role: string;
  shiftsTotal: number;
  shiftsWithViolations: number;
  averageScore: number | null;
}

export interface AnalyticsLocation {
  locationId: string;
  locationName: string;
  shiftsTotal: number;
  shiftsWithViolations: number;
  averageScore: number | null;
}

export interface AnalyticsTemplate {
  templateId: string;
  templateName: string;
  shiftsTotal: number;
  shiftsWithViolations: number;
  averageScore: number | null;
}

export interface AnalyticsCriticality {
  criticality: "critical" | "required" | "optional" | string;
  tasksTotal: number;
  done: number;
  skipped: number;
  waiverRejected: number;
  suspiciousAttachments: number;
}

export interface AnalyticsAntifake {
  attachmentsTotal: number;
  suspiciousTotal: number;
  suspiciousRate: number | null;
}

export interface AnalyticsSla {
  thresholdMin: number;
  shiftsWithActual: number;
  lateCount: number;
  lateRate: number | null;
  avgLateMin: number | null;
}

export interface AnalyticsRoleSplit {
  operator: AnalyticsKpi;
  bartender: AnalyticsKpi;
}

export interface AnalyticsDensity {
  kpis: DensityFlag;
  heatmap: DensityFlag;
  violators: DensityFlag;
  templates: DensityFlag;
}

export interface AnalyticsOverview {
  rangeFrom: string;
  rangeTo: string;
  days: number;
  kpis: AnalyticsKpi;
  heatmap: AnalyticsHeatmapCell[];
  topViolators: AnalyticsViolator[];
  locations: AnalyticsLocation[];
  templates: AnalyticsTemplate[];
  criticality: AnalyticsCriticality[];
  antifake: AnalyticsAntifake | null;
  slaLateStart: AnalyticsSla | null;
  roleSplit: AnalyticsRoleSplit | null;
  density: AnalyticsDensity | null;
  previous: AnalyticsOverview | null;
}

interface KpiDTO {
  shifts_closed: number;
  shifts_clean: number;
  shifts_with_violations: number;
  average_score: number | null;
  cleanliness_rate: number | null;
}

interface HeatmapDTO {
  day_of_week: number;
  hour_of_day: number;
  shift_count: number;
  average_score: number | null;
}

interface ViolatorDTO {
  user_id: string;
  full_name: string;
  /** Present on API v2; older servers omit it. */
  role?: string;
  shifts_total: number;
  shifts_with_violations: number;
  average_score: number | null;
}

interface LocationDTO {
  location_id: string;
  location_name: string;
  shifts_total: number;
  shifts_with_violations: number;
  average_score: number | null;
}

interface TemplateDTO {
  template_id: string;
  template_name: string;
  shifts_total: number;
  shifts_with_violations: number;
  average_score: number | null;
}

interface CriticalityDTO {
  criticality: string;
  tasks_total: number;
  done: number;
  skipped: number;
  waiver_rejected: number;
  suspicious_attachments: number;
}

interface AntifakeDTO {
  attachments_total: number;
  suspicious_total: number;
  suspicious_rate: number | null;
}

interface SlaDTO {
  threshold_min: number;
  shifts_with_actual: number;
  late_count: number;
  late_rate: number | null;
  avg_late_min: number | null;
}

interface RoleSplitDTO {
  operator: KpiDTO;
  bartender: KpiDTO;
}

interface DensityDTO {
  kpis: DensityFlag;
  heatmap: DensityFlag;
  violators: DensityFlag;
  templates: DensityFlag;
}

interface OverviewDTO {
  range_from: string;
  range_to: string;
  days: number;
  kpis: KpiDTO;
  heatmap?: HeatmapDTO[];
  top_violators?: ViolatorDTO[];
  locations?: LocationDTO[];
  /** v2 fields — absent on older `/overview` responses. */
  templates?: TemplateDTO[];
  criticality?: CriticalityDTO[];
  antifake: AntifakeDTO | null;
  sla_late_start: SlaDTO | null;
  role_split: RoleSplitDTO | null;
  density: DensityDTO | null;
  previous: OverviewDTO | null;
}

function kpiFromDto(dto: KpiDTO): AnalyticsKpi {
  return {
    shiftsClosed: dto.shifts_closed,
    shiftsClean: dto.shifts_clean,
    shiftsWithViolations: dto.shifts_with_violations,
    averageScore: dto.average_score,
    cleanlinessRate: dto.cleanliness_rate,
  };
}

function overviewFromDto(dto: OverviewDTO): AnalyticsOverview {
  // Older API deployments (or partial JSON) may omit v2 list fields — treat
  // as empty arrays so the UI never throws on `.map()` during sheet open/close
  // or after a hot reload against a stale edge.
  const heatmap = dto.heatmap ?? [];
  const topViolators = dto.top_violators ?? [];
  const locations = dto.locations ?? [];
  const templates = dto.templates ?? [];
  const criticality = dto.criticality ?? [];

  return {
    rangeFrom: dto.range_from,
    rangeTo: dto.range_to,
    days: dto.days,
    kpis: kpiFromDto(dto.kpis),
    heatmap: heatmap.map((c) => ({
      dayOfWeek: c.day_of_week,
      hourOfDay: c.hour_of_day,
      shiftCount: c.shift_count,
      averageScore: c.average_score,
    })),
    topViolators: topViolators.map((v) => ({
      userId: v.user_id,
      fullName: v.full_name,
      role: v.role ?? "",
      shiftsTotal: v.shifts_total,
      shiftsWithViolations: v.shifts_with_violations,
      averageScore: v.average_score,
    })),
    locations: locations.map((loc) => ({
      locationId: loc.location_id,
      locationName: loc.location_name,
      shiftsTotal: loc.shifts_total,
      shiftsWithViolations: loc.shifts_with_violations,
      averageScore: loc.average_score,
    })),
    templates: templates.map((t) => ({
      templateId: t.template_id,
      templateName: t.template_name,
      shiftsTotal: t.shifts_total,
      shiftsWithViolations: t.shifts_with_violations,
      averageScore: t.average_score,
    })),
    criticality: criticality.map((c) => ({
      criticality: c.criticality,
      tasksTotal: c.tasks_total,
      done: c.done,
      skipped: c.skipped,
      waiverRejected: c.waiver_rejected,
      suspiciousAttachments: c.suspicious_attachments,
    })),
    antifake: dto.antifake
      ? {
          attachmentsTotal: dto.antifake.attachments_total,
          suspiciousTotal: dto.antifake.suspicious_total,
          suspiciousRate: dto.antifake.suspicious_rate,
        }
      : null,
    slaLateStart: dto.sla_late_start
      ? {
          thresholdMin: dto.sla_late_start.threshold_min,
          shiftsWithActual: dto.sla_late_start.shifts_with_actual,
          lateCount: dto.sla_late_start.late_count,
          lateRate: dto.sla_late_start.late_rate,
          avgLateMin: dto.sla_late_start.avg_late_min,
        }
      : null,
    roleSplit: dto.role_split
      ? {
          operator: kpiFromDto(dto.role_split.operator),
          bartender: kpiFromDto(dto.role_split.bartender),
        }
      : null,
    density: dto.density
      ? {
          kpis: dto.density.kpis,
          heatmap: dto.density.heatmap,
          violators: dto.density.violators,
          templates: dto.density.templates,
        }
      : null,
    previous: dto.previous ? overviewFromDto(dto.previous) : null,
  };
}

export interface FetchOverviewParams {
  days?: number;
  /** ISO date (YYYY-MM-DD). When provided alongside `to`, overrides `days`. */
  from?: string;
  /** ISO date (YYYY-MM-DD). */
  to?: string;
  compare?: boolean;
  violatorsLimit?: number;
  locationId?: string | null;
}

export async function fetchOverview(
  params: FetchOverviewParams = {},
): Promise<ApiResult<AnalyticsOverview>> {
  const qs = new URLSearchParams();
  if (params.from && params.to) {
    qs.set("from", params.from);
    qs.set("to", params.to);
  } else if (params.days) {
    qs.set("days", String(params.days));
  }
  if (params.compare) qs.set("compare", "true");
  if (params.violatorsLimit) qs.set("violators_limit", String(params.violatorsLimit));
  if (params.locationId) qs.set("location_id", params.locationId);
  const path = `/v1/analytics/overview${qs.toString() ? `?${qs.toString()}` : ""}`;
  const result = await api.get<OverviewDTO>(path);
  if (!result.ok) return result;
  try {
    return { ok: true, status: result.status, data: overviewFromDto(result.data) };
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    return {
      ok: false,
      status: 500,
      code: "overview_parse_failed",
      message,
    };
  }
}

File: apps/api/shiftops_api/api/v1/analytics.py
"""Analytics HTTP endpoints — admin and owner only.

Why one ``/overview`` endpoint instead of one per chart
-------------------------------------------------------
The screen renders many cards from the same time window. Splitting into
per-card endpoints would multiply Supabase pooler latency for no caching
win. The aggregate query set still returns under ~120 ms on 30 days of
pilot data on the free tier (~9 SELECTs × 10 ms RTT).

Date filter
-----------
``from`` / ``to`` win over ``days`` when both are sent. They are inclusive
ISO dates (interpreted as UTC midnight). When neither is sent we fall back
to ``[now - days, now]`` so legacy callers (and the bot) keep working.
"""

from __future__ import annotations

from datetime import UTC, date, datetime, time, timedelta
from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import BaseModel
from sqlalchemy.ext.asyncio import AsyncSession

from shiftops_api.application.analytics.overview import (
    DEFAULT_DAYS,
    DEFAULT_VIOLATORS_LIMIT,
    MAX_DAYS,
    MAX_VIOLATORS_LIMIT,
    MIN_VIOLATORS_LIMIT,
    AnalyticsOverviewUseCase,
    OverviewDTO,
)
from shiftops_api.application.auth.deps import CurrentUser, require_role
from shiftops_api.domain.enums import UserRole
from shiftops_api.domain.result import Failure, Success
from shiftops_api.infra.db.engine import get_session

router = APIRouter()

_admin_or_owner = require_role(UserRole.ADMIN, UserRole.OWNER)


# --- Pydantic out-models --------------------------------------------------


class KpiBlockOut(BaseModel):
    shifts_closed: int
    shifts_clean: int
    shifts_with_violations: int
    average_score: float | None
    cleanliness_rate: float | None


class HeatmapCellOut(BaseModel):
    day_of_week: int
    hour_of_day: int
    shift_count: int
    average_score: float | None


class ViolatorRowOut(BaseModel):
    user_id: UUID
    full_name: str
    role: str
    shifts_total: int
    shifts_with_violations: int
    average_score: float | None


class LocationRowOut(BaseModel):
    location_id: UUID
    location_name: str
    shifts_total: int
    shifts_with_violations: int
    average_score: float | None


class TemplateRowOut(BaseModel):
    template_id: UUID
    template_name: str
    shifts_total: int
    shifts_with_violations: int
    average_score: float | None


class PostRowOut(BaseModel):
    location_id: UUID
    location_name: str
    slot_index: int
    station_label: str | None
    shifts_total: int
    shifts_with_violations: int
    average_score: float | None


class CriticalityRowOut(BaseModel):
    criticality: str
    tasks_total: int
    done: int
    skipped: int
    waiver_rejected: int
    suspicious_attachments: int


class AntifakeOut(BaseModel):
    attachments_total: int
    suspicious_total: int
    suspicious_rate: float | None


class SlaOut(BaseModel):
    threshold_min: int
    shifts_with_actual: int
    late_count: int
    late_rate: float | None
    avg_late_min: float | None


class RoleSplitOut(BaseModel):
    operator: KpiBlockOut
    bartender: KpiBlockOut


class DensityOut(BaseModel):
    kpis: str
    heatmap: str
    violators: str
    templates: str
    posts: str


class OverviewResponse(BaseModel):
    range_from: str
    range_to: str
    days: int
    kpis: KpiBlockOut
    heatmap: list[HeatmapCellOut]
    top_violators: list[ViolatorRowOut]
    locations: list[LocationRowOut]
    templates: list[TemplateRowOut]
    posts: list[PostRowOut]
    criticality: list[CriticalityRowOut]
    antifake: AntifakeOut | None
    sla_late_start: SlaOut | None
    role_split: RoleSplitOut | None
    density: DensityOut | None
    previous: OverviewResponse | None = None


# Allow self-recursion in the response model: previous block has same shape
# but never carries another `previous` (only the top level offers compare).
OverviewResponse.model_rebuild()


# --- Endpoint -------------------------------------------------------------


def _resolve_window(
    *,
    days: int,
    range_from: date | None,
    range_to: date | None,
) -> tuple[datetime, datetime]:
    """Convert query params into a UTC half-open ``[from, to]`` window.

    Why this lives here: the use case takes ``datetime`` objects so it can
    stay UI-agnostic. The HTTP layer is the right place to coerce date
    queries (which pilot owners type as ``2026-04-01``) into timezone-aware
    timestamps. Anyone who wants minute precision can call the use case
    directly from a script.
    """
    if range_from is not None and range_to is not None:
        if range_to < range_from:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="invalid_range",
            )
        # Inclusive end-of-day so ?to=2026-04-29 covers all of April 29th.
        rf = datetime.combine(range_from, time.min, tzinfo=UTC)
        rt = datetime.combine(range_to, time.max, tzinfo=UTC)
    else:
        rt = datetime.now(tz=UTC)
        rf = rt - timedelta(days=days)
    if (rt - rf) > timedelta(days=MAX_DAYS):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="range_too_large",
        )
    return rf, rt


def _serialise(dto: OverviewDTO) -> OverviewResponse:
    days = max(1, int((dto.range_to - dto.range_from).total_seconds() // 86400) or 1)
    return OverviewResponse(
        range_from=dto.range_from.isoformat(),
        range_to=dto.range_to.isoformat(),
        days=days,
        kpis=KpiBlockOut(
            shifts_closed=dto.kpis.shifts_closed,
            shifts_clean=dto.kpis.shifts_clean,
            shifts_with_violations=dto.kpis.shifts_with_violations,
            average_score=_to_float(dto.kpis.average_score),
            cleanliness_rate=_to_float(dto.kpis.cleanliness_rate),
        ),
        heatmap=[
            HeatmapCellOut(
                day_of_week=c.day_of_week,
                hour_of_day=c.hour_of_day,
                shift_count=c.shift_count,
                average_score=_to_float(c.average_score),
            )
            for c in dto.heatmap
        ],
        top_violators=[
            ViolatorRowOut(
                user_id=v.user_id,
                full_name=v.full_name,
                role=v.role,
                shifts_total=v.shifts_total,
                shifts_with_violations=v.shifts_with_violations,
                average_score=_to_float(v.average_score),
            )
            for v in dto.top_violators
        ],
        locations=[
            LocationRowOut(
                location_id=loc.location_id,
                location_name=loc.location_name,
                shifts_total=loc.shifts_total,
                shifts_with_violations=loc.shifts_with_violations,
                average_score=_to_float(loc.average_score),
            )
            for loc in dto.locations
        ],
        templates=[
            TemplateRowOut(
                template_id=t.template_id,
                template_name=t.template_name,
                shifts_total=t.shifts_total,
                shifts_with_violations=t.shifts_with_violations,
                average_score=_to_float(t.average_score),
            )
            for t in dto.templates
        ],
        posts=[
            PostRowOut(
                location_id=p.location_id,
                location_name=p.location_name,
                slot_index=p.slot_index,
                station_label=p.station_label,
                shifts_total=p.shifts_total,
                shifts_with_violations=p.shifts_with_violations,
                average_score=_to_float(p.average_score),
            )
            for p in dto.posts
        ],
        criticality=[
            CriticalityRowOut(
                criticality=c.criticality,
                tasks_total=c.tasks_total,
                done=c.done,
                skipped=c.skipped,
                waiver_rejected=c.waiver_rejected,
                suspicious_attachments=c.suspicious_attachments,
            )
            for c in dto.criticality
        ],
        antifake=(
            AntifakeOut(
                attachments_total=dto.antifake.attachments_total,
                suspicious_total=dto.antifake.suspicious_total,
                suspicious_rate=_to_float(dto.antifake.suspicious_rate),
            )
            if dto.antifake is not None
            else None
        ),
        sla_late_start=(
            SlaOut(
                threshold_min=dto.sla_late_start.threshold_min,
                shifts_with_actual=dto.sla_late_start.shifts_with_actual,
                late_count=dto.sla_late_start.late_count,
                late_rate=_to_float(dto.sla_late_start.late_rate),
                avg_late_min=_to_float(dto.sla_late_start.avg_late_min),
            )
            if dto.sla_late_start is not None
            else None
        ),
        role_split=(
            RoleSplitOut(
                operator=_kpi_out(dto.role_split.operator),
                bartender=_kpi_out(dto.role_split.bartender),
            )
            if dto.role_split is not None
            else None
        ),
        density=(
            DensityOut(
                kpis=dto.density.kpis,
                heatmap=dto.density.heatmap,
                violators=dto.density.violators,
                templates=dto.density.templates,
                posts=dto.density.posts,
            )
            if dto.density is not None
            else None
        ),
        previous=_serialise(dto.previous) if dto.previous is not None else None,
    )


def _to_float(v: object) -> float | None:
    if v is None:
        return None
    return float(v)  # type: ignore[arg-type]


def _kpi_out(kpi: object) -> KpiBlockOut:  # type: ignore[no-untyped-def]
    return KpiBlockOut(
        shifts_closed=kpi.shifts_closed,
        shifts_clean=kpi.shifts_clean,
        shifts_with_violations=kpi.shifts_with_violations,
        average_score=_to_float(kpi.average_score),
        cleanliness_rate=_to_float(kpi.cleanliness_rate),
    )


@router.get("/overview", response_model=OverviewResponse)
async def get_overview(
    days: Annotated[
        int,
        Query(ge=1, le=MAX_DAYS, description=f"Window in days (default {DEFAULT_DAYS})."),
    ] = DEFAULT_DAYS,
    range_from: Annotated[
        date | None,
        Query(alias="from", description="Inclusive start date (UTC), ISO format."),
    ] = None,
    range_to: Annotated[
        date | None,
        Query(alias="to", description="Inclusive end date (UTC), ISO format."),
    ] = None,
    compare: Annotated[
        bool,
        Query(description="If true, also returns the previous period of equal length."),
    ] = False,
    violators_limit: Annotated[
        int,
        Query(
            ge=MIN_VIOLATORS_LIMIT,
            le=MAX_VIOLATORS_LIMIT,
            description=f"Top N violators (default {DEFAULT_VIOLATORS_LIMIT}).",
        ),
    ] = DEFAULT_VIOLATORS_LIMIT,
    location_id: Annotated[UUID | None, Query()] = None,
    user: CurrentUser = Depends(_admin_or_owner),
    session: AsyncSession = Depends(get_session),
) -> OverviewResponse:
    rf, rt = _resolve_window(days=days, range_from=range_from, range_to=range_to)

    use_case = AnalyticsOverviewUseCase(session=session)
    result = await use_case.execute(
        user=user,
        range_from=rf,
        range_to=rt,
        location_id=location_id,
        violators_limit=violators_limit,
        compare=compare,
    )
    if isinstance(result, Failure):
        # forbidden | invalid_range | range_too_large
        if result.error.code == "forbidden":
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail=result.error.code,
            )
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=result.error.code,
        )
    assert isinstance(result, Success)
    return _serialise(result.value)
